### PR TITLE
fix: skip authentication on locked metamask

### PR DESCRIPTION
**Pull request by wa0x6e**
## 🧿 Current issues / What's wrong ?

When metamask wallet is locked *AND with a pending unlock request*, the `connect` method will return a valid provider, as if the connection was successful

This return value is wrong, and it should return nothing instead

This also means that any subsequent `connect` call on a locked metamask wallet after being denied once, will not trigger a metamask unlock call again (leaving the user figuring out by himself why the connect is not working, and that he should unlock metamask himself)

#### Reproduction

- Go to https://snapshot.box/#/
- log out and lock metamask
- log in
- do not unlock metamask and close it
- refresh the page
- Click on "Connect wallet", and choose metamask
- The button shows an avatar for `0x000` user, without username
- Clicking on the account button will open connect modal
- Clicking on metamask will do nothing
- This state will stay even after page refresh
- Unlocking metamask will auto-log the user

## 💊 Fixes / Solution

When trying to connect to a wallet in a PENDING UNLOCK state, we should trigger the unlock request again, instead of returning a valid provider

On the UI, this will:
- prevent a Schrödinger user (where the UI consider the user as logged in, even with a locked wallet), when refreshing the page and trying to connect again after denying to unlock metamask
- Trigger a metamask unlock prompt on subsequent connection (instead of just doing nothing)

## 🚧 Changes

- When trying to connect to a locked wallet pending unlocking, trigger unlock request again

## 🛠️ Tests

- `yarn build`
- `yarn link`
- Go to sx-monorepo
- `yarn link @snapshot-labs/lock` 
- `yarn dev`

#### Test 1

- Go to http://localhost:8080
- Log out and lock metamask
- Try to log in
- Ignore the metamask unlock prompt
- The "Connect wallet" button should remain in a loading state
- Open the metamask prompt manually, and unlock it
- You should be logged in with your metamask wallet

#### Test 2

- Go to http://localhost:8080
- Log out and lock metamask
- Try to log in
- Ignore the metamask unlock prompt
- The "Connect wallet" button should remain in a loading state
- Refresh the page
- The account button should be back to "Connect Wallet"
- Clicking on it will open the account modal with the connectors choice
- Clicking on "metamask" will open metamask, prompting you to unlock it

#### Test 2.a

- Unlocking metamask will log you in

#### Test 2.b 

- Do not unlock, and close the metamask popup
- The button should be back the "Connect wallet" state
- Any subsequent click, even after page refresh, should open the account modal, and connecting with metamask will open metamask asking you to unlock it

### CODE REVIEW

**Comment by ChaituVR** [comment]
Untitled 5

I get this when i retry 🤔

**Comment by wa0x6e** [comment]
I think sometime yarn link is not working.

I edited my package.json to use the local repo directly

File: connectors/injected.ts [hunk 1]
```diff
@@ -7,8 +7,20 @@ export default class Connector extends LockConnector {
...
+          try {
+            await provider.request({
+              method: "wallet_requestPermissions",
+              params: [{ eth_accounts: {} }],
+            });
+          } catch (e: any) {
+            console.error(e);
+            if (e.code === 4001) return;
+          }
+        }
```
Comment (ChaituVR) [inline connectors/injected.ts:22]: Now i got new changes in my local. but If the user ignores the second request, then it is still stuck
Untitled 7

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: connectors/injected.ts (hunk 1 above)
Comment (wa0x6e) [reply connectors/injected.ts:22]: You should not seet the second error (with requestPermission).

How many pending requests you get on the metamask icon ?

Comment (ChaituVR) [reply connectors/injected.ts:22]: 2

Comment (wa0x6e) [reply connectors/injected.ts:22]: mhhh... it should not get more than 1.

Did you log out first before locking metamask ?

Comment (ChaituVR) [reply connectors/injected.ts:22]: Yep. try this:
- Log out and lock metamask
- Try to log in
- Ignore the metamask unlock prompt
- The "Connect wallet" button should remain in a loading state
- Refresh the page
- Try to log in
- Ignore the metamask unlock prompt
- See error
- even on reload you will see same from now

Comment (wa0x6e) [reply connectors/injected.ts:22]: Did you try with sx repo on master ?

Comment (ChaituVR) [reply connectors/injected.ts:22]: Yep on master

Comment (wa0x6e) [reply connectors/injected.ts:22]: Screenshot 2024-07-27 at 19 47 53

You should get these error

**Review by ChaituVR — APPROVED** [review]
tAck

**Pull request merged**